### PR TITLE
DAOS-9755 control: Add DD_MASK and DD_SUBSYS opts to dmg set-logmasks

### PR DESCRIPTION
**Pull request by tanabarr**
DD_MASK and DD_SUBSYS values can be used to specify which debug
streams and subsystems (facilities) to enable when logging on the
server. This allows for finer grained log filtering and this change
adds support to supply extra optional arguments (--streams and
--subsystems) to dmg server set-logmasks which update the engine
process to use the new setting on relevant storage servers at runtime.
The command now takes named arguments rather than positional.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by tanabarr** [comment]
@jolivier23 should be listed as co-author for this change for help with the d_log api modifications.

**Review by knard-intel — APPROVED** [review]
LGTM

File: src/control/server/engine/config.go [hunk 1]
```diff
@@ -133,6 +136,15 @@ func NewConfig() *Config {
 	}
 }
 
+func (c *Config) ReadLogDbgStreams() (string, error) {
```
Comment (kjacque) [inline src/control/server/engine/config.go:139]: nit - needs documentation comment

File: src/control/server/engine/utils_test.go [hunk 2]
```diff
@@ -70,3 +70,42 @@ func TestValidateLogMasks(t *testing.T) {
...
+			streams: "NET",
+		},
+		"multiple streams": {
+			streams: "ANY,TRACE",
+		},
+		"unknown stream": {
+			streams: "NET,WARNING",
+			expErr:  errors.New("unknown"),
+		},
+		"mixed caae": {
```
Comment (kjacque) [inline src/control/server/engine/utils_test.go:90]: nit - typo: caae -> case

File: src/control/server/engine/utils.go [hunk 3]
```diff
@@ -98,10 +107,53 @@ func ValidateLogMasks(masks string) error {
...
 	return nil
 }
+
+var (
+	validLogStreams = []string{
+		"ALL",                                                      // Select all debug streams
+		"MD", "PL", "MGMT", "EPC", "DF", "REBUILD", "DAOS_DEFAULT", // DAOS debug streams
+		"ANY", "TRACE", "MEM", "NET", "IO", // GURT debug streams
+	}
+	errAllWithOtherLogStream = errors.New("all stream identifier cannot be used with any other")
```
Comment (kjacque) [inline src/control/server/engine/utils.go:124]: minor - might want to put "all" in quotes to ensure the meaning is clear to users

**Review by kjacque — APPROVED** [review]
Overall looks fine for me. A couple minor nits if you end up pushing again or re-running CI.

File: src/control/server/engine/config.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/engine/config.go:139]: done

File: src/control/server/engine/utils_test.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/engine/utils_test.go:90]: done

File: src/control/server/engine/utils.go (hunk 3 above)
Comment (tanabarr) [reply src/control/server/engine/utils.go:124]: done

**Review by jolivier23 — COMMENTED** [review]
Are you in process of reworking this one or should I be reviewing it as is?

**Comment by tanabarr** [comment]
> Are you in process of reworking this one or should I be reviewing it as is?

it's being reworked, I will ping when it's ready

**Comment by tanabarr** [comment]
@jolivier23 can you please have a look at the unit tests which are passing: https://github.com/daos-stack/daos/pull/12041/files#diff-3a2505d8190a5d6ac49c9ff5e09b63196f02d1b38e6011f79bf3a31ed5894102R122 and the relevant code to process the DD_X strings is https://github.com/daos-stack/daos/pull/12041/files#diff-46f0bb8a7129a8e05ff2a89566decac3f7e264be71e5f4437e1d20da8794bc12R328 , TIA.

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by jolivier23 — APPROVED** [review]
Thanks, @tanabarr.  I think this will be a nice improvement in capability

File: src/control/server/engine/utils.go [hunk 4]
```diff
@@ -51,17 +52,98 @@ const (
...
+	case LogLevelEmrg:
+		return "EMRG"
+	case LogLevelEmit:
+		return "EMIT"
+	default:
+		return ""
+	}
 }
 
+func StrToLogLevel(s string) LogLevel {
```
Comment (kjacque) [inline src/control/server/engine/utils.go:97]: nit - needs documentation comment

**Review by kjacque — APPROVED** [review]
Overall LGTM. A couple minor comments, which could be fixed in a follow-on PR.

File: src/control/lib/control/server.go [hunk 5]
```diff
@@ -38,17 +38,35 @@ func SetEngineLogMasks(ctx context.Context, rpcClient UnaryInvoker, req *SetEngi
 	if req == nil {
 		return nil, errors.New("nil request")
 	}
-	if err := engine.ValidateLogMasks(req.Masks); err != nil {
-		return nil, err
+	pbReq := new(ctlpb.SetLogMasksReq)
+
+	if req.Masks == nil {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/server.go:43]: yes think so, for example if you want to enable or disable a debug stream or subsystem/facility then you can modify just one of the options

File: src/control/server/engine/utils.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/engine/utils.go:97]: will address on the next push or in a follow-up PR

File: src/control/cmd/dmg/server.go [hunk 6]
```diff
@@ -28,12 +28,9 @@ type serverSetLogMasksCmd struct {
...
+	Subsystems *string `short:"s" long:"subsystems" description:"This input string is equivalent to the use of the DD_SUBSYS environment variable and can be set to enable logging for specific subsystems or facilities. The expected syntax is a comma separated list of facility identifiers. Accepted DAOS facilitiesm are common,tree,vos,client,server,rdb,pool,container,object,placement,rebuild,tier,mgmt,bio,tests, Common facilities (GURT) are MISC,MEM and CaRT facilities RPC,BULK,CORPC,GRP,LM,HG,ST,IV If not set, subsystems to enable will be read from server config file and if set to an empty string then logging all subsystems will be enabled"`
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/dmg/server.go:33]: will address on the next push or in a follow-up PR

[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/cmd/dmg/server.go:33]: done

File: src/control/server/engine/utils.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/engine/utils.go:97]: done

**Review by knard-intel — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]

**Comment by tanabarr** [comment]
@daos-stack/daos-gatekeeper all relevant checks have passed, Clang Format check is failing for a reason unrelated to this PR. Please can we land?

**Comment by ashleypittman** [comment]
> @daos-stack/daos-gatekeeper all relevant checks have passed, Clang Format check is failing for a reason unrelated to this PR. Please can we land?

clang-format is new and only advisory, the warnings did relate to this PR but it also needs updating to ignore the protobuf files.

**Pull request merged**